### ISSUE TITLE
Problem with `viewer` variable with multiple viewers and consolas

### ISSUE THREAD

**Issue by Czaki** [napari/napari-console#40]
We inject `viewer` variable into the console to allow a user to interact with the viewer.

However, if we have multiple viewers in a single process, the `viewer` variable will point to last viewer where we try to open the console (even if the second console do not show https://github.com/napari/napari/issues/7413, that will be fixed in #39) 

The variable is defined here:

https://github.com/napari/napari-console/blob/463f828d5f4af732dfd87a7091d2ec4c8647af17/napari_console/qt_console.py#L106

It will be nice to find that there is already `viewer` variable defined in console and add `viewer1` etc. 

@dalthviz could you take a look?

**Comment by psobolewskiPhD** [comment]
I'm confused by the concepts here.
I tested a simple script:
```
import napari
v1 = napari.Viewer()
v2 = napari.Viewer()
napari.run()
```
and I was quite shocked when neither v1 nor v2 worked in either console.
But `viewer` worked in both and used the proper viewer, that associated with each console I was using--which I didn't expect based on this issue.
So to me the issue isn't how `viewer` works, but rather that the `v1` and `v2` I put in my script arn't pushed to the console(s) -- obviously I couldn't use `viewer` twice!

Other than a script the other way to get in a 2 viewer shared state is to launch napari from the terminal, which has a console, and use that console to make another viewer.
I did this with `v2 = napari.Viewer()` and got a second viewer with a console working. Surprisingly, `v2` worked correctly there! But so did `viewer` -- in the 2nd viewer, either variable worked!
I made a `data` array and it was available in both consoles.
But `viewer.add_image` added it only to the 2nd viewer, regardless which viewer I used. So is *this* the nature of the bug?

**Comment by Czaki** [comment]
After you open console in second viewer, did you try to interact with first viewer via console?

**Comment by jni** [comment]
Yeah like @psobolewskiPhD, I don't understand why #18 didn't result in v1 and v2 being injected into the namespace... I think fixing that is more important than the workaround in #41 (though thank you @dalthviz!), and indeed I would remove the special-casing around the viewer — the viewer should have whatever variable name the script gave it.

**Comment by psobolewskiPhD** [comment]
> After you open console in second viewer, did you try to interact with first viewer via console?

OK i retested this and indeed and it's quite strange!
If I start using v1 console, i can add and interact with that viewer no problem.
If I switch to v2 console, viewer.add_image adds to v2. OK. But after this, everything in the v1 console returns v2 info, e.g. `v1.title`.
If from the beginning I use `viewer` in the `v2` console, then the v1 console always interacts with v2.

**Comment by Czaki** [comment]
@dalthviz Maybe it will be possible to overwrite `viewer` variable when gain focus to console?

**Comment by psobolewskiPhD** [comment]
If I did `v1 = napari.Viewer()` I wouldn't expect `viewer` to also work.
Using a simple case:
```
import napari
v1 = napari.Viewer()
napari.run()
```
When I open the console of *that* viewer, `v1` doesn't work with `NameError: name 'v1' is not defined` but `viewer` does.
I think this is very unintuitive and I consider this a bug.

**Comment by dalthviz** [comment]
> When I open the console of that viewer, v1 doesn't work with NameError: name 'v1' is not defined

Oh so for you the variable in the script is not working? Checking the simple case with only one viewer I'm able to access the viewer via `v1` and `viewer` when running the console from main, in fact if I do `v1 == viewer` I get `True`:

![imagen](https://github.com/user-attachments/assets/4a13dfc3-2e18-4198-931f-96af6246c329)

Maybe the variable is working for my due to the way I'm running the script (from a cmd running `python script.py`)? 🤔 Are you running the script in a somehow different way @psobolewskiPhD ?

**Comment by psobolewskiPhD** [comment]
I'm also using `python test.py`
I have:
```
Name: napari-console
Version: 0.1.1
```
and napari from main.

**Comment by dalthviz** [comment]
I'm being unable to reproduce that behavior :/ (for me the variables over the script work) From what I understand #18 should have enabled the use of script variables in the console, right? Maybe there are some setups/cases where the implementation done there doesn't work? :thinking: Could it be worthy to open a new issue for that then?

**Comment by psobolewskiPhD** [comment]
I just did a fresh env, `pip install "napari[all]"`
I got napari-console 0.1.2 and I tested
```
import napari
my_var = "tada!"
v1 = napari.Viewer()
napari.run()
```
This works, I can access v1 and my_var from the console. viewer as well.

**Comment by psobolewskiPhD** [comment]
I just retested in the new env with
```
import napari
my_var = "tada!"
v1 = napari.Viewer(title="v1")
v2 = napari.Viewer(title="v2")
napari.run()
```
I can access v1 and v2 in either console and both work just fine.
SO it's just `viewer` that is odd: here is the v2 console:
```
In [3]: viewer.title        # v2 console
Out[3]: 'v2'
```
```
In [4]: viewer.title          # switched to v1 console
Out[4]: 'v1'
```
Back to v2 console:
```
viewer.add_image(data2)
Out[5]: <Image layer 'data2' at 0x16e3cbfe0>
```
viewer.title
Out[6]: 'v1'
```

**Comment by dalthviz** [comment]
> I can access v1 and v2 in either console and both work just fine.

Awesome! 🎉 

>SO it's just viewer that is odd

I opened PR #42 which should take care of that. There the `viewer` variable gets updated to point to the viewer of the current console being used (idea pointed out at https://github.com/napari/napari-console/issues/40#issuecomment-2525322961). Let me know if that approach and if the interaction with the variables previewed there makes sense!

**Comment by psobolewskiPhD** [comment]
I made a fresh dev env: 
`pip install -e .[dev, pyqt6]`
and v1 and v2 don't work...

but with `pip install napari[all]` it works.
with `pip install -e .[all]` it also works!
with `pip install -e .[pyqt6]` it *doesnt work*. So looks like some pyqt6 compatibility? or am I going crazy?

**Comment by dalthviz** [comment]
I was able to reproduce the missing variables when using PyQt6! From a quick check, seems like adding `qtpy.` to the `_PREF_LIST` constant (which is used over the `_in_napari` method and the logic to capture variables) helps. Will add a commit with the change over #42

**Comment by psobolewskiPhD** [comment]
Crosslinking: https://github.com/napari/napari/issues/1530 : clobbering `viewer` with multiple consoles.
Also this comment from Talley: https://github.com/napari/napari/issues/787#issuecomment-564829450 check if `viewer` is in globals and don't clobber.
